File: tests/src/com.google.protobuf/protobuf-java/2.5.0/src/test/java/com_google_protobuf/protobuf_java/GeneratedMessageLiteInnerSerializedFormTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_google_protobuf.protobuf_java;

import com.google.protobuf.DescriptorProtos;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class GeneratedMessageLiteInnerSerializedFormTest {
    @Test
    void deserializesGeneratedMessageThroughSerializedFormReadResolve() throws Exception {
        DescriptorProtos.DescriptorProto messageType = DescriptorProtos.DescriptorProto.newBuilder()
                .setName("Person")
                .build();
        DescriptorProtos.FileDescriptorProto original = DescriptorProtos.FileDescriptorProto.newBuilder()
                .setName("sample.proto")
                .setPackage("sample")
                .addDependency("common.proto")
                .addMessageType(messageType)
                .build();

        Object restored = deserialize(serialize(original));

        assertThat(restored).isInstanceOf(DescriptorProtos.FileDescriptorProto.class);
        assertThat(restored).isNotSameAs(original).isEqualTo(original);
        assertThat(((DescriptorProtos.FileDescriptorProto) restored).toByteArray()).isEqualTo(original.toByteArray());
    }

    private static byte[] serialize(Object value) throws Exception {
        ByteArrayOutputStream bytes = new ByteArrayOutputStream();
        try (ObjectOutputStream output = new ObjectOutputStream(bytes)) {
            output.writeObject(value);
        }
        return bytes.toByteArray();
    }

    private static Object deserialize(byte[] bytes) throws Exception {
        try (ObjectInputStream input = new ObjectInputStream(new ByteArrayInputStream(bytes))) {
            return input.readObject();
        }
    }
}

File: tests/src/org.jetbrains.kotlinx/kotlinx-metadata-jvm/0.9.0/src/test/kotlin/org_jetbrains_kotlinx/kotlinx_metadata_jvm/GeneratedMessageLiteInnerSerializedFormTest.kt
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_jetbrains_kotlinx.kotlinx_metadata_jvm

import kotlinx.metadata.internal.metadata.ProtoBuf
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

public class GeneratedMessageLiteInnerSerializedFormTest {
    @Test
    public fun javaSerializationRestoresGeneratedMessageLiteMessages(): Unit {
        val original: ProtoBuf.StringTable = ProtoBuf.StringTable.newBuilder()
            .addString("kotlinx")
            .addString("metadata")
            .build()

        val restored: Any = deserialize(serialize(original))

        assertThat(restored).isInstanceOf(ProtoBuf.StringTable::class.java)
        val restoredStringTable: ProtoBuf.StringTable = restored as ProtoBuf.StringTable
        assertThat(restoredStringTable.stringList).containsExactly("kotlinx", "metadata")
        assertThat(restoredStringTable.toByteArray()).isEqualTo(original.toByteArray())
    }

    private fun serialize(value: Any): ByteArray {
        val byteOutput: ByteArrayOutputStream = ByteArrayOutputStream()
        ObjectOutputStream(byteOutput).use { objectOutput: ObjectOutputStream ->
            objectOutput.writeObject(value)
        }
        return byteOutput.toByteArray()
    }

    private fun deserialize(bytes: ByteArray): Any {
        return ObjectInputStream(ByteArrayInputStream(bytes)).use { objectInput: ObjectInputStream ->
            objectInput.readObject()
        }
    }
}
